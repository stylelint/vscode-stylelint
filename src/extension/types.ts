import type TypedEventEmitter from 'typed-emitter';
import type { DidRegisterDocumentFormattingEditProviderNotificationParams } from '../server/index';

/**
 * VS Code extension event names.
 */
export type ExtensionEvents = {
	DidRegisterDocumentFormattingEditProvider: (
		params: DidRegisterDocumentFormattingEditProviderNotificationParams,
	) => void;
	DidResetConfiguration: () => void;
};

/**
 * VS Code extension public API.
 */
export type PublicApi = TypedEventEmitter<ExtensionEvents> & {
	codeActionReady: boolean;
};

/**
 * Extension API event names.
 */
export enum ApiEvent {
	DidRegisterDocumentFormattingEditProvider = 'DidRegisterDocumentFormattingEditProvider',
	DidResetConfiguration = 'DidResetConfiguration',
}
