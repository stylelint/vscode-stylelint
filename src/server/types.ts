import type { Connection, Disposable } from 'vscode-languageserver';
import type { TextDocuments } from 'vscode-languageserver/node';
import type { TextDocument, TextEdit } from 'vscode-languageserver-textdocument';
import { CodeActionKind as VSCodeActionKind } from 'vscode-languageserver-types';
import type LSP from 'vscode-languageserver-protocol';
// eslint-disable-next-line n/no-unpublished-import
import type stylelint from 'stylelint';
import type winston from 'winston';
import type { StylelintRunner, LintDiagnostics } from '../utils/stylelint/index';
import type { ExtractKeysOfValueType } from '../utils/types';
import type { PackageManager, StylelintResolutionResult } from '../utils/packages/index';
import type { NotificationManager, CommandManager } from '../utils/lsp/index';

/**
 * Command IDs
 */
export enum CommandId {
	ApplyAutoFix = 'stylelint.applyAutoFix',
	OpenRuleDoc = 'stylelint.openRuleDoc',
}

/**
 * Code action kinds
 */
export const CodeActionKind = {
	StylelintSourceFixAll: `${VSCodeActionKind.SourceFixAll}.stylelint`,
};

/**
 * Language server notification names.
 */
export enum Notification {
	DidRegisterCodeActionRequestHandler = 'stylelint/didRegisterCodeActionRequestHandler',
	DidRegisterDocumentFormattingEditProvider = 'stylelint/didRegisterDocumentFormattingEditProvider',
	DidResetConfiguration = 'stylelint/didResetConfiguration',
}

/**
 * `DidRegisterDocumentFormattingEditProvider` notification parameters.
 */
export interface DidRegisterDocumentFormattingEditProviderNotificationParams {
	/**
	 * The URI of the document for which the formatting provider was registered.
	 */
	readonly uri: string;

	/**
	 * The options used to register the document formatting provider.
	 */
	readonly options: LSP.DocumentFormattingRegistrationOptions;
}

/**
 * Context shared between the language server and its modules.
 */
export interface LanguageServerContext {
	/**
	 * The language server connection.
	 */
	connection: Connection;

	/**
	 * The notification manager for the connection.
	 */
	notifications: NotificationManager;

	/**
	 * The command manager for the connection.
	 */
	commands: CommandManager;

	/**
	 * The text document manager.
	 */
	documents: TextDocuments<TextDocument>;

	/**
	 * The runner with which to run Stylelint.
	 */
	runner: StylelintRunner;

	/**
	 * Displays the given error in the client using the language server
	 * connection.
	 * @param error The error to display.
	 */
	displayError(error: unknown): void;

	/**
	 * Gets the effective extension options for a resource, given its URI.
	 * @param uri The resource URI.
	 */
	getOptions(uri: string): Promise<LanguageServerOptions>;

	/**
	 * Returns the module with the given ID if it exists.
	 * @param id The ID of the module to return.
	 */
	getModule(id: string): LanguageServerModule | undefined;

	/**
	 * Lints a document using Stylelint and returns fix text edits.
	 * @param document The document to get text edits for.
	 * @param linterOptions Options to pass to the linter. Overridden by the
	 * language server options.
	 */
	getFixes(document: TextDocument, linterOptions?: stylelint.LinterOptions): Promise<TextEdit[]>;

	/**
	 * Lints a document using Stylelint and returns diagnostics.
	 * @param document The document to lint.
	 * @param linterOptions Options to pass to the linter. Overridden by the
	 * language server options.
	 */
	lintDocument(
		document: TextDocument,
		linterOptions?: Partial<stylelint.LinterOptions>,
	): Promise<LintDiagnostics | undefined>;

	/**
	 * Resolves the Stylelint package to be used for the given document.
	 * @param document The document to resolve the package for.
	 */
	resolveStylelint(document: TextDocument): Promise<StylelintResolutionResult | undefined>;
}

/**
 * A language server module.
 */
export interface LanguageServerModule extends Disposable {
	/**
	 * Handler called when the server is initializing.
	 */
	onInitialize?: (params: LSP.InitializeParams) => Partial<LSP.InitializeResult> | undefined | void;

	/**
	 * Handler called after the language server finishes registering handlers
	 * with the connection.
	 */
	onDidRegisterHandlers?: () => void;

	/**
	 * Handler called after the language server has finished responding to the
	 * onDidChangeConfiguration event.
	 */
	onDidChangeConfiguration?: () => Promise<void>;

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	[key: string | symbol]: any;
}

/**
 * A language server module class.
 */
export interface LanguageServerModuleConstructor {
	/**
	 * The module's ID, used to identify the module in the language server's
	 * internal state and when logging. Should be a unique, short, lowercase
	 * string.
	 */
	id: string;
	new (params: LanguageServerModuleConstructorParameters): LanguageServerModule;
}

/**
 * Parameters for the {@link LanguageServerModuleConstructor} constructor.
 */
export type LanguageServerModuleConstructorParameters = {
	context: LanguageServerContext;
	logger?: winston.Logger;
};

/**
 * Language server event handler names.
 */
export type LanguageServerHandlers = ExtractKeysOfValueType<LanguageServerModule, () => unknown>;

/**
 * Parameters for language server event handlers, keyed by the handler name.
 */
export type LanguageServerHandlerParameters = {
	[key in LanguageServerHandlers]: Parameters<Required<LanguageServerModule>[key]>;
};

/**
 * Return types for language server event handlers, keyed by the handler name.
 */
export type LanguageServerHandlerReturnValues = {
	[key in LanguageServerHandlers]: ReturnType<Required<LanguageServerModule>[key]>;
};

/**
 * Language server constructor parameters.
 */
export type LanguageServerConstructorParameters = {
	/**
	 * The language server connection.
	 */
	connection: Connection;

	/**
	 * The logger to use. If not provided, a default logger will be used.
	 */
	logger?: winston.Logger;

	/**
	 * The modules to load.
	 */
	modules?: LanguageServerModuleConstructor[];
};

/**
 * Language server options.
 */
export type LanguageServerOptions = {
	codeAction: {
		disableRuleComment: {
			location: 'separateLine' | 'sameLine';
		};
	};
	config?: stylelint.Config | null;
	configBasedir?: string;
	configFile?: string;
	customSyntax?: string;
	ignoreDisables?: boolean;
	packageManager: PackageManager;
	reportDescriptionlessDisables?: boolean;
	reportInvalidScopeDisables?: boolean;
	reportNeedlessDisables?: boolean;
	snippet: string[];
	stylelintPath?: string;
	validate: string[];
};
