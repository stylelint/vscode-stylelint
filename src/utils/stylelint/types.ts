import type LSP from 'vscode-languageserver-protocol';
// eslint-disable-next-line n/no-unpublished-import
import type stylelint from 'stylelint';
import type { PackageManager } from '../packages/index';
export type Stylelint = typeof stylelint;
export type ConfigurationError = Error & { code: 78 };

/**
 * Diagnostics for a lint run.
 */
export type LintDiagnostics = {
	/**
	 * The diagnostics, each corresponding to a warning or error emitted by
	 * Stylelint.
	 */
	diagnostics: LSP.Diagnostic[];

	/**
	 * Raw output from Stylelint, if any.
	 */
	output?: string;

	/**
	 * Gets the original warning from the given diagnostic.
	 */
	getWarning?: (diagnostic: LSP.Diagnostic) => stylelint.Warning | null;
};

/**
 * Disable report rule names.
 */
export enum DisableReportRuleNames {
	Needless = '--report-needless-disables',
	InvalidScope = '--report-invalid-scope-disables',
	Descriptionless = '--report-descriptionless-disables',
	Illegal = 'reportDisables',
}

/**
 * Stylelint runner options.
 */
export type RunnerOptions = {
	config?: stylelint.Config | null;
	configBasedir?: string;
	configFile?: string;
	customSyntax?: string;
	ignoreDisables?: boolean;
	packageManager?: PackageManager;
	reportDescriptionlessDisables?: boolean;
	reportInvalidScopeDisables?: boolean;
	reportNeedlessDisables?: boolean;
	snippet?: string[];
	stylelintPath?: string;
	validate?: string[];
};

/**
 * Error thrown when a rule's option is invalid.
 */
export class InvalidOptionError extends Error {
	reasons: string[];

	constructor(warnings: { text: string }[]) {
		const reasons = warnings.map((warning) => warning.text);

		super(reasons.join('\n'));
		this.reasons = reasons;
	}
}
