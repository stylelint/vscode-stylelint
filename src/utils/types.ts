/**
 * Makes all properties of type `T` optional except for those in `K`.
 */
export type OptionalExcept<T extends Record<never, never>, K extends keyof T> = Pick<T, K> &
	Partial<Pick<T, K>>;

/**
 * Makes all properties of type `T` required except for those in `K`.
 */
export type RequiredExcept<T extends Record<never, never>, K extends keyof T> = Pick<
	T,
	Exclude<keyof T, K>
> &
	Required<Pick<T, K>>;

/**
 * Extracts keys from a type `T` the values of which are type `V`.
 */
export type ExtractKeysOfValueType<T, V> = {
	[K in keyof T]-?: T[K] extends V ? K : never;
}[keyof T];

/**
 * Given a function `T`, returns a function type with the same parameters, but
 * with the return type `Promise<R> | R` where `R` is the return type of `T`.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type MaybeAsync<T extends (...args: any[]) => any> = (
	...args: Parameters<T>
) => Promise<ReturnType<T>> | ReturnType<T>;

/**
 * Given a type `T`, returns a type with only the public properties of `T`.
 */
export type PublicOnly<T> = Pick<T, keyof T>;

/**
 * Given a type `T`, returns a type with only the public properties of `T` and
 * its descendants.
 */
export type PublicOnlyDeep<T> = {
	// eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
	[K in keyof T]: T[K] extends Function ? T[K] : T[K] extends object ? PublicOnlyDeep<T[K]> : T[K];
};
