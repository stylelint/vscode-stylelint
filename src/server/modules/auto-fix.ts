import type { TextDocument } from 'vscode-languageserver-textdocument';
import { WorkspaceChange } from 'vscode-languageserver-protocol';
import type LSP from 'vscode-languageserver-protocol';
import { CommandId } from '../types';
import type {
	LanguageServerContext,
	LanguageServerModuleConstructorParameters,
	LanguageServerModule,
} from '../types';
import type winston from 'winston';

export class AutoFixModule implements LanguageServerModule {
	static id = 'auto-fix';

	/**
	 * The language server context.
	 */
	#context: LanguageServerContext;

	/**
	 * The logger to use, if any.
	 */
	#logger: winston.Logger | undefined;

	/**
	 * Disposables for handlers.
	 */
	#disposables: LSP.Disposable[] = [];

	constructor({ context, logger }: LanguageServerModuleConstructorParameters) {
		this.#context = context;
		this.#logger = logger;
	}

	async #shouldAutoFix(document: TextDocument): Promise<boolean> {
		const options = await this.#context.getOptions(document.uri);

		return options.validate.includes(document.languageId);
	}

	onInitialize(): Partial<LSP.InitializeResult> {
		return {
			capabilities: {
				executeCommandProvider: {
					commands: [CommandId.ApplyAutoFix],
				},
			},
		};
	}

	dispose(): void {
		this.#disposables.forEach((disposable) => disposable.dispose());
		this.#disposables.length = 0;
	}

	onDidRegisterHandlers(): void {
		this.#logger?.debug('Registering onExecuteCommand handler');

		this.#disposables.push(
			this.#context.commands.on(CommandId.ApplyAutoFix, async ({ arguments: args }) => {
				if (!args) {
					return {};
				}

				const identifier = args[0] as { version: number; uri: string };
				const uri = identifier.uri;
				const document = this.#context.documents.get(uri);

				if (!document || !(await this.#shouldAutoFix(document))) {
					if (this.#logger?.isDebugEnabled()) {
						if (!document) {
							this.#logger.debug('Unknown document, ignoring', { uri });
						} else {
							this.#logger.debug('Document should not be auto-fixed, ignoring', {
								uri,
								language: document.languageId,
							});
						}
					}

					return {};
				}

				if (identifier.version !== document.version) {
					this.#logger?.debug('Document has been modified, ignoring', {
						uri,
					});

					return {};
				}

				const workspaceChange = new WorkspaceChange();
				const textChange = workspaceChange.getTextEditChange(identifier);

				const edits = await this.#context.getFixes(document);

				edits.forEach((edit) => textChange.add(edit));

				this.#logger?.debug('Applying fixes', { uri, edits });

				try {
					const response = await this.#context.connection.workspace.applyEdit(workspaceChange.edit);

					if (!response.applied) {
						this.#logger?.debug('Failed to apply fixes', { uri, response });
					}
				} catch (error) {
					this.#logger?.debug('Failed to apply fixes', { uri, error });
				}

				return {};
			}),
		);
	}
}
