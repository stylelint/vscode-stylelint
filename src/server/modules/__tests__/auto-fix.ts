import { TextDocument } from 'vscode-languageserver-textdocument';
import { Position, TextEdit } from 'vscode-languageserver-types';
import type LSP from 'vscode-languageserver-protocol';
import { CommandId } from '../../types';

import { AutoFixModule } from '../auto-fix';
import { WorkDoneProgressReporter } from 'vscode-languageserver';

const mockContext = serverMocks.getContext();
const mockLogger = serverMocks.getLogger();

describe('AutoFixModule', () => {
	beforeEach(() => {
		mockContext.__options.validate = [];
		jest.clearAllMocks();
	});

	test('should be constructable', () => {
		expect(() => new AutoFixModule({ context: mockContext.__typed() })).not.toThrow();
	});

	test('onInitialize should return results', () => {
		const module = new AutoFixModule({ context: mockContext.__typed() });

		expect(module.onInitialize()).toMatchSnapshot();
	});

	test('onDidRegisterHandlers should register an auto-fix command handler', () => {
		const module = new AutoFixModule({ context: mockContext.__typed() });

		module.onDidRegisterHandlers();

		expect(mockContext.commands.on).toHaveBeenCalledTimes(1);
		expect(mockContext.commands.on).toHaveBeenCalledWith(
			CommandId.ApplyAutoFix,
			expect.any(Function),
		);
	});

	test('should auto-fix documents', async () => {
		const document = TextDocument.create('foo', 'bar', 1, 'a {}');

		mockContext.documents.get.mockReturnValue(document);
		mockContext.__options.validate = ['bar'];
		mockContext.getFixes.mockResolvedValue([TextEdit.insert(Position.create(0, 0), 'text')]);
		mockContext.connection.workspace.applyEdit.mockResolvedValue({
			applied: true,
		});

		const module = new AutoFixModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.commands.on.mock.calls.find(
			([command]) => command === CommandId.ApplyAutoFix,
		)?.[1];

		const result = await handler?.(
			{
				command: CommandId.ApplyAutoFix,
				arguments: [{ uri: 'foo', version: 1 }],
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toStrictEqual({});
		expect(mockContext.getFixes).toHaveBeenCalledWith(document);
		expect(mockContext.connection.workspace.applyEdit.mock.calls[0]).toMatchSnapshot();
	});

	test('with incorrect command, should not attempt to auto-fix', async () => {
		const module = new AutoFixModule({ context: mockContext.__typed() });

		module.onDidRegisterHandlers();

		const handler = mockContext.commands.on.mock.calls.find(
			([command]) => command === CommandId.ApplyAutoFix,
		)?.[1];

		const result = await handler?.(
			{ command: 'foo' },
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toStrictEqual({});
		expect(mockContext.getFixes).not.toHaveBeenCalled();
	});

	test('with no arguments, should not attempt to auto-fix', async () => {
		const module = new AutoFixModule({ context: mockContext.__typed() });

		module.onDidRegisterHandlers();

		const handler = mockContext.commands.on.mock.calls.find(
			([command]) => command === CommandId.ApplyAutoFix,
		)?.[1];

		const result = await handler?.(
			{ command: CommandId.ApplyAutoFix },
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toStrictEqual({});
		expect(mockContext.getFixes).not.toHaveBeenCalled();
	});

	test('if no matching document exists, should not attempt to auto-fix', async () => {
		mockContext.documents.get.mockReturnValue(undefined);

		const module = new AutoFixModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.commands.on.mock.calls.find(
			([command]) => command === CommandId.ApplyAutoFix,
		)?.[1];

		const result = await handler?.(
			{
				command: CommandId.ApplyAutoFix,
				arguments: [{ uri: 'foo' }],
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toStrictEqual({});
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Unknown document, ignoring', { uri: 'foo' });
	});

	test('if document language ID is not in options, should not attempt to auto-fix', async () => {
		mockContext.documents.get.mockReturnValue(TextDocument.create('foo', 'bar', 1, 'a {}'));
		mockContext.__options.validate = ['baz'];

		const module = new AutoFixModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.commands.on.mock.calls.find(
			([command]) => command === CommandId.ApplyAutoFix,
		)?.[1];

		const result = await handler?.(
			{
				command: CommandId.ApplyAutoFix,
				arguments: [{ uri: 'foo' }],
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toStrictEqual({});
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith(
			'Document should not be auto-fixed, ignoring',
			{
				uri: 'foo',
				language: 'bar',
			},
		);
	});

	test('with no debug log level and no valid document, should not attempt to log reason', async () => {
		mockContext.documents.get.mockReturnValue(TextDocument.create('foo', 'bar', 1, 'a {}'));
		mockContext.__options.validate = ['baz'];
		mockLogger.isDebugEnabled.mockReturnValue(false);

		const module = new AutoFixModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.commands.on.mock.calls.find(
			([command]) => command === CommandId.ApplyAutoFix,
		)?.[1];

		const handlerParams = {
			command: CommandId.ApplyAutoFix,
			arguments: [{ uri: 'foo' }],
		};

		const result = await handler?.(
			handlerParams,
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toStrictEqual({});
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Registering onExecuteCommand handler');
	});

	test('if the document has been modified, should not attempt to auto-fix', async () => {
		mockContext.documents.get.mockReturnValue(TextDocument.create('foo', 'bar', 2, 'a {}'));
		mockContext.__options.validate = ['bar'];

		const module = new AutoFixModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.commands.on.mock.calls.find(
			([command]) => command === CommandId.ApplyAutoFix,
		)?.[1];

		const result = await handler?.(
			{
				command: CommandId.ApplyAutoFix,
				arguments: [{ uri: 'foo', version: 1 }],
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toStrictEqual({});
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Document has been modified, ignoring', {
			uri: 'foo',
		});
	});

	test('if fixes fail to apply, should log the response', async () => {
		const response = { applied: false, failureReason: 'foo' };

		mockContext.documents.get.mockReturnValue(TextDocument.create('foo', 'bar', 1, 'a {}'));
		mockContext.__options.validate = ['bar'];
		mockContext.getFixes.mockResolvedValue([TextEdit.insert(Position.create(0, 0), 'text')]);
		mockContext.connection.workspace.applyEdit.mockResolvedValue(response);

		const module = new AutoFixModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.commands.on.mock.calls.find(
			([command]) => command === CommandId.ApplyAutoFix,
		)?.[1];

		const result = await handler?.(
			{
				command: CommandId.ApplyAutoFix,
				arguments: [{ uri: 'foo', version: 1 }],
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toStrictEqual({});
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Failed to apply fixes', {
			uri: 'foo',
			response,
		});
	});

	test('if an error occurs while applying fixes, should log it', async () => {
		const error = new Error('foo');

		mockContext.documents.get.mockReturnValue(TextDocument.create('foo', 'bar', 1, 'a {}'));
		mockContext.__options.validate = ['bar'];
		mockContext.getFixes.mockResolvedValue([TextEdit.insert(Position.create(0, 0), 'text')]);
		mockContext.connection.workspace.applyEdit.mockRejectedValue(error);

		const module = new AutoFixModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.commands.on.mock.calls.find(
			([command]) => command === CommandId.ApplyAutoFix,
		)?.[1];

		const result = await handler?.(
			{
				command: CommandId.ApplyAutoFix,
				arguments: [{ uri: 'foo', version: 1 }],
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toStrictEqual({});
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Failed to apply fixes', {
			uri: 'foo',
			error,
		});
	});

	it('should be disposable', () => {
		const module = new AutoFixModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		expect(module).toHaveProperty('dispose');
		expect(module.dispose).toBeInstanceOf(Function);
	});

	it('should dispose all handler registrations when disposed', () => {
		const module = new AutoFixModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();
		module.dispose();

		const disposable = mockContext.commands.on.mock.results[0].value;

		expect(disposable.dispose).toHaveBeenCalledTimes(1);
	});
});
