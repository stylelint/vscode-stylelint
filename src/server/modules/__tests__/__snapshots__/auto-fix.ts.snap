// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoFixModule onInitialize should return results 1`] = `
{
  "capabilities": {
    "executeCommandProvider": {
      "commands": [
        "stylelint.applyAutoFix",
      ],
    },
  },
}
`;

exports[`AutoFixModule should auto-fix documents 1`] = `
[
  {
    "changeAnnotations": undefined,
    "documentChanges": [
      {
        "edits": [
          {
            "newText": "text",
            "range": {
              "end": {
                "character": 0,
                "line": 0,
              },
              "start": {
                "character": 0,
                "line": 0,
              },
            },
          },
        ],
        "textDocument": {
          "uri": "foo",
          "version": 1,
        },
      },
    ],
  },
]
`;
